### ISSUE TITLE
Pluggable JWT Authenticator in Presto

### ISSUE THREAD

**Issue by imjalpreet** [prestodb/presto#24052]
# Background

JWT (JSON Web Token) authentication in Presto is used to ensure secure access to the Presto cluster by validating the identity of the clients requesting resources. 

Presto can be configured to authenticate incoming requests using JWTs. This works by attaching a token to the request, which Presto verifies before granting access.

## Setup Process

- Enable JWT authentication by configuring the authentication.type to JWT.
- You’ll need to configure the `http-server.authentication.jwt.key-file`, which points to the public key used to verify the JWT’s signature.
- Optionally, `http-server.authentication.jwt.required-issuer` and `http-server.authentication.jwt.required-audience` can be configured to enforce validation on specific fields within the token.

## Client-Side Usage

When requesting Presto, the client includes the JWT in the Authorization header as a Bearer token:

> Authorization: Bearer <token>
> 

## Configuration Example

Here’s a basic configuration setup in Presto’s `config.properties`:

```
http-server.authentication.type=JWT
http-server.authentication.jwt.key-file=/path/to/public/key.pem
http-server.authentication.jwt.required-issuer=trusted-issuer
http-server.authentication.jwt.required-audience=your-presto-cluster
```

`key-file`: Path to the public key (PEM format) used to validate the JWT’s signature.

`required-issuer`: Specifies the trusted issuer of the JWT (e.g., an identity provider like Auth0).

`required-audience`: Ensures the token is intended for the Presto cluster.

# Motivation

A pluggable authenticator in Presto allows the authentication process to be customized based on specific use cases, such as integrating with different identity providers or token validation strategies. This document focuses on JWT, a widely used token format that encapsulates authentication information.

Presto already supports authentication with a username and password via a custom password authenticator that validates the credentials and creates a principal.

As part of this effort, we would like to add support for authenticating via a custom JWT Authenticator that validates the token and creates a principal.

# Custom JWT Authenticator Implementation

The current generic implementation for `JsonWebTokenAuthenticator` was moved back to Presto from Airlift (package: `http-server`) a few months back.

Three main classes that are part of the current generic implementation are:

- `com.facebook.presto.server.security.JsonWebTokenAuthenticator`
- `com.facebook.presto.server.security.JsonWebTokenConfig`
- `com.facebook.airlift.http.server.BasicPrincipal`

 

## Introduce JWTAuthenticatorFactory in Presto SPI

`JWTAuthenticatorFactory` is responsible for creating a `JWTAuthenticator` instance. It also defines the name of this authenticator which is used by the administrator in a Presto configuration.

```java
String getName();

JWTAuthenticator create(Map<String, String> config);
```

## Introduce JWTAuthenticator in Presto SPI

`JWTAuthenticator` contains a single method, `createAuthenticatedPrincipal()`, that validates the token and returns a `Principal`, which is then authorized by the [System Access Control](https://prestodb.github.io/docs/0.289/develop/system-access-control.html).

The implementation of `JWTAuthenticatorFactory` must be wrapped as a plugin and installed on the Presto cluster.

```java
/**
 * Authenticate the provided token.
 *
 * @return the authenticated Principal
 * @throws AccessDeniedException if not allowed
 */
Principal createAuthenticatedPrincipal(String token, HttpServletRequest request);
```

## Introduce JWTAuthenticatorManager in Presto Main

`JWTAuthenticatorManager` is responsible for managing the registration and loading of JWT Authenticator Plugins. 

After a plugin that implements `JWTAuthenticatorFactory` has been installed on the coordinator, it is configured using an `etc/jwt-authenticator.properties` file. All of the properties other than `jwt-authenticator.name` is specific to the  `JWTAuthenticatorFactory` implementation.

The `jwt-authenticator.name` property is used by Presto to find a registered  `JWTAuthenticatorFactory` based on the name returned by  `JWTAuthenticatorFactory.getName()`. 

The remaining properties are passed as a map to `JWTAuthenticatorFactory.create()` by the `JWTAuthenticationManager`.

## Refactor current `JsonWebTokenAuthenticator` implementation

Refactor the generic implementation of `JsonWebTokenAuthenticator` to JWTAuthenticator Plugin.

This will be the default implementation of `JWTAuthenticator` and `JWTAuthenticatorFactory`.

**Comment by tdcmeehan** [comment]
Have you considered the SPI being an `Authenticator`, rather than a `JwtAuthenticator`?

**Comment by imjalpreet** [comment]
@tdcmeehan At this stage, the decision to introduce a JwtAuthenticator SPI was made to address the specific requirements of JWT-based authentication, as we already have a similar SPI for password-based authentication. Initially, I thought it might be better to first implement pluggable JWT authenticators and, as a follow-up, work on designing a more generic Authenticator SPI since it would introduce a breaking change for everyone who has already developed custom password authenticators.

However, thank you for raising this point. I had also given it a thought in the past. I agree that unifying both PasswordAuthenticator and JwtAuthenticator into a single, generic Authenticator SPI is the right approach for the long term. This would provide a consistent interface for all authentication mechanisms and simplify integration and extensibility.

The proposed Authenticator SPI would define a common method, such as:

```
Principal createAuthenticatedPrincipal(Map<String, String> credentials, HttpServletRequest request);
```

This approach allows flexibility in handling different types of credentials:
	•	For JWT, the Map could have: { "token": "<jwt_token>" }
	•	For username/password, it could have: { "user": "<user>", "password": "<pass>" }

Both the existing PasswordAuthenticator and the new JwtAuthenticator can be refactored to implement this interface without altering their core functionality, ensuring backward compatibility while improving the overall design.

I’d love to hear your thoughts on how we should proceed and the design of a generic Authenticator SPI.

**Comment by tdcmeehan** [comment]
@imjalpreet this sounds good, but actually I believe all the required parameters would be found in the `request`, specifically the headers, so I don't think we would need an additional `credentials` map.  Just pass in the headers and let the authenticator self-selected which ones may be used for authentication.

**Comment by imjalpreet** [comment]
@tdcmeehan I agree all necessary parameters can be extracted from the request. I would like to clarify why I recommended using the map.

In the current implementation of the PasswordAuthenticator SPI, we directly pass the user credentials along with the password. The inclusion of the `HttpServletRequest request` in the JWTAuthenticator SPI was necessary due to a recent change made for JWT Authentication, which can be found here: https://github.com/prestodb/presto/commit/fd2615e89562047daa004851015270007c6de746. This change requires us to update the original request based on the token received.

I believe it would be beneficial to parse the credentials and include them in the SPI. This approach could help eliminate the boilerplate code that would otherwise be necessary in each custom authenticator implementation. For example, in the case of Password Authenticators (whether it’s File, LDAP, or any custom authenticator used by Presto users), we would need to include the following code in every authenticator implementation: https://github.com/prestodb/presto/blob/1b88d3b11a0dba073c8b3d4efc3fdb88ee6d7247/presto-main/src/main/java/com/facebook/presto/server/security/PasswordAuthenticator.java#L49-L64
This change could also present a significant breaking change, as anyone utilizing their custom implementations—rather than the File or LDAP options provided out-of-the-box in Presto—would now need to handle request parsing.

Let me know your thoughts on the same.

**Comment by imjalpreet** [comment]
There has been a slight change since my last comment on this issue. After discussions with @tdcmeehan, we introduced a new and more generic SPI that is independent of the current Authenticators. This SPI can be used to implement any type of custom authenticator, without being limited to just a custom implementation of JWT or Password Authenticators.

We decided to keep the current implementation of Password and JWT authenticators unchanged to avoid introducing a breaking change.

The proposed Authenticator SPI would define a common method, such as:

`Principal createAuthenticatedPrincipal(HttpServletRequest request);`
